File: app/src/main/java/com/giaquino/android/cropped/ui/login/LoginViewModel.java
package com.giaquino.android.cropped.ui.login;

import androidx.hilt.Assisted;
import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import com.giaquino.android.cropped.common.Constant;
import com.giaquino.android.cropped.data.api.request.AccessTokenRequest;
import com.giaquino.android.cropped.data.common.Resource;
import com.giaquino.android.cropped.data.model.AccessToken;
import com.giaquino.android.cropped.data.repository.SharedRepository;
import com.giaquino.android.cropped.data.repository.UserRepository;

public class LoginViewModel extends ViewModel {

    private UserRepository userRepository;
    private SharedRepository sharedRepository;
    private MediatorLiveData<Resource<AccessToken>> accessToken = new MediatorLiveData<>();


    @ViewModelInject
    public LoginViewModel(UserRepository userRepository, SharedRepository sharedRepository) {
        this.userRepository = userRepository;
        this.sharedRepository = sharedRepository;
    }

    public boolean isLoggedIn() {
        return sharedRepository.isLoggedIn();
    }

    public LiveData<Resource<AccessToken>> accessToken() {
        return accessToken;
    }

    public void authorize(String code) {
        AccessTokenRequest request = new AccessTokenRequest(
                Constant.UNSPLASH_CLIENT_ID,
                Constant.UNSPLASH_CLIENT_SECRET,
                Constant.UNSPLASH_REDIRECT_URI,
                code);

        LiveData<Resource<AccessToken>> liveData = LiveDataReactiveStreams.fromPublisher(
                userRepository.authorize(request).toFlowable());

        accessToken.setValue(Resource.loading(null));

        accessToken.addSource(liveData, resource -> {
            accessToken.setValue(resource);
            accessToken.removeSource(liveData);
        });
    }
}

File: app/src/main/java/com/giaquino/android/cropped/ui/login/LoginViewModel.kt
package com.giaquino.android.cropped.ui.login

import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import com.giaquino.android.cropped.data.api.request.TokenRequest
import com.giaquino.android.cropped.data.common.Resource
import com.giaquino.android.cropped.data.model.Token
import com.giaquino.android.cropped.data.repository.SharedRepository
import com.giaquino.android.cropped.data.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(
        private val userRepository: UserRepository,
        private val sharedRepository: SharedRepository) : ViewModel() {

    private val _token = MediatorLiveData<Resource<Token>>()
    val token: LiveData<Resource<Token>> = _token

    fun isLoggedIn() : Boolean {
        return sharedRepository.getUnsplashToken() != null
    }

    fun authorize(code: String) {
        val source: LiveData<Resource<Token>> = LiveDataReactiveStreams
                .fromPublisher(userRepository.authorize(TokenRequest(code = code)).toFlowable())

        _token.apply {
            postValue(Resource.loading(null))
            addSource(source) {
                postValue(it)
                removeSource(source)
            }
        }
    }
}